micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,448 @@
+}
+
+/// StaticMemoryDataStore is an entity that governs the static bytes memory
+/// reserved in DATA_STORE. The memory is segmented based according to the DataCollectors expected
+/// (identified by collector keys) and each segment view is treated as a separate ring buffer of SensorData
+/// messages. Currently, an equal amount of space is alloted to each collector, which will affect
+/// the maximum allowed size of a single message (see get_maximum_message_size below)
+/// It should be treated as a global struct that should only be initialized once and is not
+/// thread-safe (all interactions should be blocking).
+pub struct StaticMemoryDataStore {
+    buffers: Vec<LocalRb<u8, &'static mut [MaybeUninit<u8>]>>,
+    collector_keys: Vec<ResourceMethodKey>,
+}
+
+impl StaticMemoryDataStore {
+    pub fn new(collector_keys: Vec<ResourceMethodKey>) -> Result<Self, DataStoreError> {
+        unsafe {
+            if !DATA_STORE_INITIALIZED.fetch_or(true, Ordering::SeqCst) {
+                let len_per_buffer = DATA_STORE.len() / collector_keys.len();
+                let mut buffers = Vec::new();
+                for i in 0..(collector_keys.len()) {
+                    let start_idx = i * len_per_buffer;
+                    let end_idx = (i + 1) * len_per_buffer;
+                    buffers.push(LocalRb::from_raw_parts(
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+                return Ok(Self {
+                    buffers,
+                    collector_keys,
+                });
+            }
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == &collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key))
+    }
+
+    fn get_maximum_message_size(&self) -> usize {
+        let len_per_buffer = unsafe { DATA_STORE.len() / self.collector_keys.len() };
+        len_per_buffer * 2
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_messages(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        messages: Vec<SensorData>,
+    ) -> Result<Vec<SensorData>, DataStoreError> {
+        let max_msg_size = self.get_maximum_message_size();
+        let buffer_index = self.get_index_for_collector(collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let mut res = Vec::new();
+        let mut return_remaining = false;
+        for req in messages {
+            if return_remaining {
+                res.push(req);
+                continue;
+            }
+            let encode_len = req.encoded_len();
+            if encode_len > max_msg_size {
+                return Err(DataStoreError::DataTooLarge);
```

Comment:
This is sort of the scenario I was mentioning above: if you return this error, you may have already written some of the SensorData from `messages`, but you don't get the benefit of the `Vec` return.